### ISSUE TITLE
DatasetScans for S3

### ISSUE THREAD

**Issue by tdrwenski** [Unidata/tds#358]
- Get DatasetScans working for S3.
- Add tests for S3 including: scanning a bucket, scanning a "directory", filtering, sorting
- Add documentation examples
See also https://github.com/Unidata/tds/issues/271

**Comment by WeatherGod** [comment]
As motivation, consider NOAA's Open Data bucket: `s3://noaa-gfs-bdp-pds`. Unfortunately, they didn't split different data types into separate prefixes at the top-level, so filtering is going to be critical to be able to serve out only certain datasets from this bucket. For example, I'm trying to serve up 00z GFS model runs formatted `gfs.#yyyyMMdd/00/atmos/gfs.t00z.pgrb2.0p25.f*`. I have tried doing it as an aggregation with a regEx to do a "filtering", but I gave up after an hour of TDS trying to return the ncml of the dataset.

The only way we've been able to make this even remotely work is to put an AWS File Storage Gateway in front of the bucket and then mount the share in our TDS instance and utilize a DatasetScan, but the performance is horrible due to the storage gateway.
